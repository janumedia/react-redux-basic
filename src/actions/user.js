import { ADD_USER, DELETE_USER } from "../constants/action-type";

export const addUser = (newUser) => ({
    type: ADD_USER,
    payload: newUser
})

export const deleteUser = (userID) => ({
    type: DELETE_USER,
    payload: userID
})
