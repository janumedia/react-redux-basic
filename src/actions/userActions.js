import * as types from "../constants/action-types";

export const addUser = (newUser) => ({
    type: types.ADD_USER,
    payload: newUser
})

export const deleteUser = (userID) => ({
    type: types.DELETE_USER,
    payload: userID
})

export const getUsers = () => ({
    type: types.GET_USERS,
    payload: {url: '/data/users.json'}
})
