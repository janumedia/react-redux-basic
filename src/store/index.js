import { createStore, applyMiddleware } from 'redux';

import userReducer from '../reducers/userReducer';
import userMiddleWare from '../middleware/userMiddleWare';

export default createStore (userReducer, applyMiddleware(userMiddleWare));