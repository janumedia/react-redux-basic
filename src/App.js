import React from 'react';
import { connect } from 'react-redux';

import { getUsers} from './actions/userActions';

import Header from './components/Header';
import Quote from './components/Quote';
import List from './components/List';
import Form from './components/Form';
import Status from './components/Status';

const mapStateToProps = state => state;

const mapDispatchToProps = dispatch => {
    return {
        getUsers: () => dispatch(getUsers())
    };
};

class App extends React.Component {
    componentDidMount() {
        this.props.getUsers()
    }
    render() {
        return (
            <React.Fragment>
                <Header/>
                <section>
                    <Quote>User List</Quote>
                    <Form>
                        <Status/>
                    </Form>
                    <List users={this.props.users}/>
                </section>
            </React.Fragment>
        );
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(App)