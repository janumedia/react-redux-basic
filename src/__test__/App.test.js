import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';

import axios from 'axios';

import App from '../App';
import store from '../store/index';
import * as userActions from '../actions/userActions';

jest.mock('axios');

jest.useFakeTimers();

describe('App component', () => {

    const user = {
        id: 4,
        name: 'Dave Mustaine',
        role: 'Engineer'
    }

    const user2 = {
        id: 6,
        name: 'Tony Stark',
        role: 'CEO'
    }

    const testAndSimulateForm = async (component, name, role) => {
        let tree = component.toJSON();
    
        const sectionItems = tree.filter(node => node.type === 'section')[0].children;
        const formDiv = sectionItems.find(node => node.props.className === 'Form').children;
        const form = formDiv.find(node => node.type === 'form');
        const inputList = form.children[0].children.filter(node => node.type === 'input');
        const nameInput = inputList[0];
        const roleInput = inputList[1];
        
        //simulate input onChange
        nameInput.props.onChange({target:{value:name}});
        roleInput.props.onChange({target:{value:role}});
    
        //initial
        tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    
        form.props.onSubmit({preventDefault:()=>{}});
    
        //update onSubmiting
        tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    
        jest.runOnlyPendingTimers();
        const mockAddUser = addUserSpy.mockResolvedValueOnce();
        await mockAddUser();
    
        //update after onSubmited
        tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    }

    let component, getUsersSpy, addUserSpy, deleteUserSpy;

    beforeEach(() => {
        
        getUsersSpy = jest.spyOn(userActions, 'getUsers');
        addUserSpy = jest.spyOn(userActions, 'addUser');
        deleteUserSpy = jest.spyOn(userActions, 'deleteUser');
       
        axios.get.mockResolvedValueOnce({
            data:
            {
                users:[user]
            }
        });

        component = renderer.create(
            <Provider store={store}>
                <App/>
            </Provider>
        )
    });

    afterAll(() => {
        getUsersSpy.mockRestore();
        addUserSpy.mockRestore();
        deleteUserSpy.mockRestore();
        jest.clearAllTimers();
    })

    it('should render emty user list initially', () => {
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should render list of user items', async () => {
        
        //initial
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();

        const mockGetUsers = getUsersSpy.mockResolvedValueOnce();
        jest.runOnlyPendingTimers();
        await mockGetUsers();
        
        //update
        tree = component.toJSON();
        expect(tree).toMatchSnapshot();

    });

    it('should add user item onSubmit form', async () => {
        
        const mockGetUsers = getUsersSpy.mockResolvedValueOnce();
        jest.runOnlyPendingTimers();
        await mockGetUsers();
        
        testAndSimulateForm(component, user2.name, user2.role);
        
    });

    it('should show error status when submit existing name', async () => {
        
        const mockGetUsers = getUsersSpy.mockResolvedValueOnce();
        jest.runOnlyPendingTimers();
        await mockGetUsers();
        
        testAndSimulateForm(component, user.name, user.role);
        
    });

    it('should remove user item delete button clicked', async () => {
        
        jest.runOnlyPendingTimers();
        const mockGetUsers = getUsersSpy.mockResolvedValueOnce();
        await mockGetUsers();
        
        //initial
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
        
        const sectionItems = tree.filter(node => node.type === 'section')[0].children;
        const listUser = sectionItems.filter(node => node.type === 'ul')[0].children;
        const itemUser = listUser[0].children;
        const deleteBtn = itemUser[1].children.find(node => node.type === 'button');
        //simulate onClick
        deleteBtn.props.onClick();

        //update on deleting
        tree = component.toJSON();
        expect(tree).toMatchSnapshot();   

        jest.runOnlyPendingTimers();
        const mockDeleteUser = deleteUserSpy.mockResolvedValueOnce();
        await mockDeleteUser()

        //update after deleted
        tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

})