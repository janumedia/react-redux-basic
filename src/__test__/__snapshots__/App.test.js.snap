// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should add user item onSubmit form 1`] = `
Array [
  <div
    className="Header"
  >
    <ul>
      <li>
        <a
          href="/"
        >
          Home
        </a>
      </li>
      <li>
        <a
          href="about"
        >
          About
        </a>
      </li>
      <li>
        <a
          href="works"
        >
          Works
        </a>
      </li>
    </ul>
  </div>,
  <section>
    <h1>
      User List
    </h1>
    <div
      className="Form"
    >
      <form
        onSubmit={[Function]}
      >
        <div>
          <label>
            Name
          </label>
          <input
            id="name"
            onChange={[Function]}
            type="text"
            value="Tony Stark"
          />
          <label>
            Role
          </label>
          <input
            id="role"
            onChange={[Function]}
            type="text"
            value="CEO"
          />
        </div>
        <button
          type="submit"
        >
          Add User
        </button>
      </form>
      <p
        className="Status "
      >
        
      </p>
    </div>
    <ul>
      <div
        className="User  "
      >
        <div>
          <p>
            Name: 
            Dave Mustaine
          </p>
          <p>
            Role: 
            Engineer
          </p>
        </div>
        <div>
          <button
            onClick={[Function]}
          >
            Delete
          </button>
        </div>
      </div>
    </ul>
  </section>,
]
`;

exports[`App component should add user item onSubmit form 2`] = `
Array [
  <div
    className="Header"
  >
    <ul>
      <li>
        <a
          href="/"
        >
          Home
        </a>
      </li>
      <li>
        <a
          href="about"
        >
          About
        </a>
      </li>
      <li>
        <a
          href="works"
        >
          Works
        </a>
      </li>
    </ul>
  </div>,
  <section>
    <h1>
      User List
    </h1>
    <div
      className="Form"
    >
      <form
        onSubmit={[Function]}
      >
        <div>
          <label>
            Name
          </label>
          <input
            id="name"
            onChange={[Function]}
            type="text"
            value="Tony Stark"
          />
          <label>
            Role
          </label>
          <input
            id="role"
            onChange={[Function]}
            type="text"
            value="CEO"
          />
        </div>
        <button
          type="submit"
        >
          Add User
        </button>
      </form>
      <p
        className="Status pending"
      >
        please wait...!
      </p>
    </div>
    <ul>
      <div
        className="User  "
      >
        <div>
          <p>
            Name: 
            Dave Mustaine
          </p>
          <p>
            Role: 
            Engineer
          </p>
        </div>
        <div>
          <button
            onClick={[Function]}
          >
            Delete
          </button>
        </div>
      </div>
    </ul>
  </section>,
]
`;

exports[`App component should add user item onSubmit form 3`] = `
Array [
  <div
    className="Header"
  >
    <ul>
      <li>
        <a
          href="/"
        >
          Home
        </a>
      </li>
      <li>
        <a
          href="about"
        >
          About
        </a>
      </li>
      <li>
        <a
          href="works"
        >
          Works
        </a>
      </li>
    </ul>
  </div>,
  <section>
    <h1>
      User List
    </h1>
    <div
      className="Form"
    >
      <form
        onSubmit={[Function]}
      >
        <div>
          <label>
            Name
          </label>
          <input
            id="name"
            onChange={[Function]}
            type="text"
            value=""
          />
          <label>
            Role
          </label>
          <input
            id="role"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <button
          type="submit"
        >
          Add User
        </button>
      </form>
      <p
        className="Status "
      >
        
      </p>
    </div>
    <ul>
      <div
        className="User hidden "
      >
        <div>
          <p>
            Name: 
            Tony Stark
          </p>
          <p>
            Role: 
            CEO
          </p>
        </div>
        <div>
          <button
            onClick={[Function]}
          >
            Delete
          </button>
        </div>
      </div>
      <div
        className="User  "
      >
        <div>
          <p>
            Name: 
            Dave Mustaine
          </p>
          <p>
            Role: 
            Engineer
          </p>
        </div>
        <div>
          <button
            onClick={[Function]}
          >
            Delete
          </button>
        </div>
      </div>
    </ul>
  </section>,
]
`;

exports[`App component should remove user item delete button clicked 1`] = `
Array [
  <div
    className="Header"
  >
    <ul>
      <li>
        <a
          href="/"
        >
          Home
        </a>
      </li>
      <li>
        <a
          href="about"
        >
          About
        </a>
      </li>
      <li>
        <a
          href="works"
        >
          Works
        </a>
      </li>
    </ul>
  </div>,
  <section>
    <h1>
      User List
    </h1>
    <div
      className="Form"
    >
      <form
        onSubmit={[Function]}
      >
        <div>
          <label>
            Name
          </label>
          <input
            id="name"
            onChange={[Function]}
            type="text"
            value=""
          />
          <label>
            Role
          </label>
          <input
            id="role"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <button
          type="submit"
        >
          Add User
        </button>
      </form>
      <p
        className="Status "
      >
        
      </p>
    </div>
    <ul>
      <div
        className="User  "
      >
        <div>
          <p>
            Name: 
            Dave Mustaine
          </p>
          <p>
            Role: 
            Engineer
          </p>
        </div>
        <div>
          <button
            onClick={[Function]}
          >
            Delete
          </button>
        </div>
      </div>
    </ul>
  </section>,
]
`;

exports[`App component should remove user item delete button clicked 2`] = `
Array [
  <div
    className="Header"
  >
    <ul>
      <li>
        <a
          href="/"
        >
          Home
        </a>
      </li>
      <li>
        <a
          href="about"
        >
          About
        </a>
      </li>
      <li>
        <a
          href="works"
        >
          Works
        </a>
      </li>
    </ul>
  </div>,
  <section>
    <h1>
      User List
    </h1>
    <div
      className="Form"
    >
      <form
        onSubmit={[Function]}
      >
        <div>
          <label>
            Name
          </label>
          <input
            id="name"
            onChange={[Function]}
            type="text"
            value=""
          />
          <label>
            Role
          </label>
          <input
            id="role"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <button
          type="submit"
        >
          Add User
        </button>
      </form>
      <p
        className="Status "
      >
        
      </p>
    </div>
    <ul>
      <div
        className="User  deleting"
      >
        <div>
          <p>
            Name: 
            Dave Mustaine
          </p>
          <p>
            Role: 
            Engineer
          </p>
        </div>
        <div>
          <button
            onClick={[Function]}
          >
            Deleting
          </button>
        </div>
      </div>
    </ul>
  </section>,
]
`;

exports[`App component should remove user item delete button clicked 3`] = `
Array [
  <div
    className="Header"
  >
    <ul>
      <li>
        <a
          href="/"
        >
          Home
        </a>
      </li>
      <li>
        <a
          href="about"
        >
          About
        </a>
      </li>
      <li>
        <a
          href="works"
        >
          Works
        </a>
      </li>
    </ul>
  </div>,
  <section>
    <h1>
      User List
    </h1>
    <div
      className="Form"
    >
      <form
        onSubmit={[Function]}
      >
        <div>
          <label>
            Name
          </label>
          <input
            id="name"
            onChange={[Function]}
            type="text"
            value=""
          />
          <label>
            Role
          </label>
          <input
            id="role"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <button
          type="submit"
        >
          Add User
        </button>
      </form>
      <p
        className="Status "
      >
        
      </p>
    </div>
    <ul />
  </section>,
]
`;

exports[`App component should render emty user list initially 1`] = `
Array [
  <div
    className="Header"
  >
    <ul>
      <li>
        <a
          href="/"
        >
          Home
        </a>
      </li>
      <li>
        <a
          href="about"
        >
          About
        </a>
      </li>
      <li>
        <a
          href="works"
        >
          Works
        </a>
      </li>
    </ul>
  </div>,
  <section>
    <h1>
      User List
    </h1>
    <div
      className="Form"
    >
      <form
        onSubmit={[Function]}
      >
        <div>
          <label>
            Name
          </label>
          <input
            id="name"
            onChange={[Function]}
            type="text"
            value=""
          />
          <label>
            Role
          </label>
          <input
            id="role"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <button
          type="submit"
        >
          Add User
        </button>
      </form>
      <p
        className="Status pending"
      >
        please wait...!
      </p>
    </div>
    <ul />
  </section>,
]
`;

exports[`App component should render list of user items 1`] = `
Array [
  <div
    className="Header"
  >
    <ul>
      <li>
        <a
          href="/"
        >
          Home
        </a>
      </li>
      <li>
        <a
          href="about"
        >
          About
        </a>
      </li>
      <li>
        <a
          href="works"
        >
          Works
        </a>
      </li>
    </ul>
  </div>,
  <section>
    <h1>
      User List
    </h1>
    <div
      className="Form"
    >
      <form
        onSubmit={[Function]}
      >
        <div>
          <label>
            Name
          </label>
          <input
            id="name"
            onChange={[Function]}
            type="text"
            value=""
          />
          <label>
            Role
          </label>
          <input
            id="role"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <button
          type="submit"
        >
          Add User
        </button>
      </form>
      <p
        className="Status pending"
      >
        please wait...!
      </p>
    </div>
    <ul />
  </section>,
]
`;

exports[`App component should render list of user items 2`] = `
Array [
  <div
    className="Header"
  >
    <ul>
      <li>
        <a
          href="/"
        >
          Home
        </a>
      </li>
      <li>
        <a
          href="about"
        >
          About
        </a>
      </li>
      <li>
        <a
          href="works"
        >
          Works
        </a>
      </li>
    </ul>
  </div>,
  <section>
    <h1>
      User List
    </h1>
    <div
      className="Form"
    >
      <form
        onSubmit={[Function]}
      >
        <div>
          <label>
            Name
          </label>
          <input
            id="name"
            onChange={[Function]}
            type="text"
            value=""
          />
          <label>
            Role
          </label>
          <input
            id="role"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <button
          type="submit"
        >
          Add User
        </button>
      </form>
      <p
        className="Status "
      >
        
      </p>
    </div>
    <ul>
      <div
        className="User hidden "
      >
        <div>
          <p>
            Name: 
            Dave Mustaine
          </p>
          <p>
            Role: 
            Engineer
          </p>
        </div>
        <div>
          <button
            onClick={[Function]}
          >
            Delete
          </button>
        </div>
      </div>
    </ul>
  </section>,
]
`;

exports[`App component should show error status when submit existing name 1`] = `
Array [
  <div
    className="Header"
  >
    <ul>
      <li>
        <a
          href="/"
        >
          Home
        </a>
      </li>
      <li>
        <a
          href="about"
        >
          About
        </a>
      </li>
      <li>
        <a
          href="works"
        >
          Works
        </a>
      </li>
    </ul>
  </div>,
  <section>
    <h1>
      User List
    </h1>
    <div
      className="Form"
    >
      <form
        onSubmit={[Function]}
      >
        <div>
          <label>
            Name
          </label>
          <input
            id="name"
            onChange={[Function]}
            type="text"
            value="Dave Mustaine"
          />
          <label>
            Role
          </label>
          <input
            id="role"
            onChange={[Function]}
            type="text"
            value="Engineer"
          />
        </div>
        <button
          type="submit"
        >
          Add User
        </button>
      </form>
      <p
        className="Status "
      >
        
      </p>
    </div>
    <ul>
      <div
        className="User  "
      >
        <div>
          <p>
            Name: 
            Dave Mustaine
          </p>
          <p>
            Role: 
            Engineer
          </p>
        </div>
        <div>
          <button
            onClick={[Function]}
          >
            Delete
          </button>
        </div>
      </div>
    </ul>
  </section>,
]
`;

exports[`App component should show error status when submit existing name 2`] = `
Array [
  <div
    className="Header"
  >
    <ul>
      <li>
        <a
          href="/"
        >
          Home
        </a>
      </li>
      <li>
        <a
          href="about"
        >
          About
        </a>
      </li>
      <li>
        <a
          href="works"
        >
          Works
        </a>
      </li>
    </ul>
  </div>,
  <section>
    <h1>
      User List
    </h1>
    <div
      className="Form"
    >
      <form
        onSubmit={[Function]}
      >
        <div>
          <label>
            Name
          </label>
          <input
            id="name"
            onChange={[Function]}
            type="text"
            value="Dave Mustaine"
          />
          <label>
            Role
          </label>
          <input
            id="role"
            onChange={[Function]}
            type="text"
            value="Engineer"
          />
        </div>
        <button
          type="submit"
        >
          Add User
        </button>
      </form>
      <p
        className="Status pending"
      >
        please wait...!
      </p>
    </div>
    <ul>
      <div
        className="User  "
      >
        <div>
          <p>
            Name: 
            Dave Mustaine
          </p>
          <p>
            Role: 
            Engineer
          </p>
        </div>
        <div>
          <button
            onClick={[Function]}
          >
            Delete
          </button>
        </div>
      </div>
    </ul>
  </section>,
]
`;

exports[`App component should show error status when submit existing name 3`] = `
Array [
  <div
    className="Header"
  >
    <ul>
      <li>
        <a
          href="/"
        >
          Home
        </a>
      </li>
      <li>
        <a
          href="about"
        >
          About
        </a>
      </li>
      <li>
        <a
          href="works"
        >
          Works
        </a>
      </li>
    </ul>
  </div>,
  <section>
    <h1>
      User List
    </h1>
    <div
      className="Form"
    >
      <form
        onSubmit={[Function]}
      >
        <div>
          <label>
            Name
          </label>
          <input
            id="name"
            onChange={[Function]}
            type="text"
            value="Dave Mustaine"
          />
          <label>
            Role
          </label>
          <input
            id="role"
            onChange={[Function]}
            type="text"
            value="Engineer"
          />
        </div>
        <button
          type="submit"
        >
          Add User
        </button>
      </form>
      <p
        className="Status error"
      >
        Dave Mustaine is already exist!
      </p>
    </div>
    <ul>
      <div
        className="User  "
      >
        <div>
          <p>
            Name: 
            Dave Mustaine
          </p>
          <p>
            Role: 
            Engineer
          </p>
        </div>
        <div>
          <button
            onClick={[Function]}
          >
            Delete
          </button>
        </div>
      </div>
    </ul>
  </section>,
]
`;
